### PR TITLE
Use authenticated user in graphql responses

### PR DESCRIPTION
**Pull request by LucasPickering**
This adds the logged-in user to the graphql context, and updates all the user program API handlers to reference that user instead of the `tmp_user` hack. I had to add a bunch of tests to account for new edge cases that occur when the user isn't logged in, or tries to access another user's programs.

Currently this breaks some shit on the frontend if you're not logged in (cause it tries to fetch the user programs but that gives an error). I removed one of these cases but the others i left in place. I'm gonna work on that soon.

### CODE REVIEW

File: api/src/server/gql/mutation.rs [hunk 1]
```diff
@@ -239,14 +240,19 @@ impl MutationFields for Mutation {
...
-                .optional();
+        // Update the row, returning the new value. If the row doesn't
+        // exist, this will return None.
+        let result: Result<Option<models::UserProgram>, _> = diesel::update(
+            user_programs::table
+                .find(modified_user_program.id)
+                // Extra filter to make sure the requesting user is the owner of
+                // the row.
+                .filter(user_programs::columns::user_id.eq(user.id)),
+        )
```
Comment (JRMurr) [inline api/src/server/gql/mutation.rs:251]: Should we throw an error saying you dont have access to this program instead of doing nothing?

Comment (LucasPickering) [reply api/src/server/gql/mutation.rs:251]: I think in general, if a user doesn't have read perms on an object, then any time they try to do something to it we should pretend like it doesn't exist, to prevent leaking any data. In this case it wouldn't be a big deal cause we would just be leaking that a certain guid exists in the DB, but i think it's still the best practice.

**Comment by LucasPickering** [comment]
@JRMurr There were some import conflicts on this after merging the other PR so I fixed those, didn't make any changes otherwise tho.

File: api/src/server/gql/mutation.rs (hunk 1 above)
Comment (JRMurr) [reply api/src/server/gql/mutation.rs:251]: image
